### ISSUE TITLE
Crash caused by other programs' dbus activity

### ISSUE THREAD

**Issue by calvin-sykes** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#3048]
**OS:** Ubuntu 16.04 LTS

**GPMDP Version:** 4.4.1, 4.5.0

**Issue Descriptions:** 

Emacs' AUCTeX mode supports 'inverse search' (clicking in the PDF takes you to the line in the LaTeX source and vice versa) which uses DBus behind the scenes.
If GPMDP is running when this feature is used (whether music is playing or not), it immediately crashes.

**Steps to Reproduce:**

prerequisites:

- either GPMDP 4.4.5 or 4.5.0
- Emacs 24.5.1
- Ubuntu 16.04 "Document Viewer" (Evince 3.18.2)
- TeX Live

steps:

- Open GPMDP
- In Emacs, create a file containing the following LaTeX:
```
\documentclass{minimal}
\usepackage{lipsum}
\begin{document}
\lipsum[1]
\end{document}
```
- Save (Ctrl-X Ctrl-S) and enable SyncTeX (Ctrl-C Ctrl-T Ctrl-S)
- Compile and open the PDF in Evince (Ctrl-C Ctrl-A)
- Attempt to use inverse search by Ctrl-clicking in the PDF

Within a few seconds GPMDP should abruptly close.
If music is playing it will stutter once before the crash.

I've attached a (filtered) `dbus-monitor` log resulting from doing the above.
This shows the reverse search message as well as the GrabMediaPlayerKeys member which I'd guess is how GPMDP is affected.
Those messages still appear when reverse search isn't used though (and no crash occurs).

[dbus.log](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/files/1809076/dbus.log)

Running GPMDP from the command line gives the following (unfortunately less-than-helpful!) output when it crashes:

```
Failed to get crash dump id.
Report Id: 
Segmentation fault (core dumped)
```

**Comment by jostrander** [comment]
Wow, that's a first. Maybe related to #2864, with the parameters being backwards?

**Comment by jostrander** [comment]
What are you using to get that dbus log?

**Comment by calvin-sykes** [comment]
Seems like it - the patch in that issue fixes this.

**Comment by jostrander** [comment]
Care to PR it in?

**Comment by calvin-sykes** [comment]
After opening a PR I found a typo in the patch which meant it was just disabling media key support completely.

Fixing that brought the issue back so I closed the PR.
Provisionally, using the `dbus-native` package instead of `dbus` is a successful workaround.

Patch for that:
```diff
diff --git a/src/main/features/linux/mediaKeysDBus.js b/src/main/features/linux/mediaKeysDBus.js
index 87deaaf..b9461ce 100644
--- a/src/main/features/linux/mediaKeysDBus.js
+++ b/src/main/features/linux/mediaKeysDBus.js
@@ -1,30 +1,52 @@
-import DBus from 'dbus';
+import { sessionBus } from 'dbus-native';
 
-function registerBindings(desktopEnv, session) {
-  session.getInterface(`org.${desktopEnv}.SettingsDaemon`,
-  `/org/${desktopEnv}/SettingsDaemon/MediaKeys`,
-  `org.${desktopEnv}.SettingsDaemon.MediaKeys`, (err, iface) => {
-    if (!err) {
-      iface.on('MediaPlayerKeyPressed', (n, keyName) => {
-        switch (keyName) {
-          case 'Next': Emitter.sendToGooglePlayMusic('playback:nextTrack'); return;
-          case 'Previous': Emitter.sendToGooglePlayMusic('playback:previousTrack'); return;
-          case 'Play': Emitter.sendToGooglePlayMusic('playback:playPause'); return;
-          case 'Stop': Emitter.sendToGooglePlayMusic('playback:stop'); return;
-          default: return;
-        }
-      });
-      iface.GrabMediaPlayerKeys(0, `org.${desktopEnv}.SettingsDaemon.MediaKeys`); // eslint-disable-line
+function registerBindings(service, session) {
+  session.getService(service.bus_name).getInterface(
+    service.path,
+    service.interface,
+    (err, iface) => {
+      if (!err) {
+        iface.on('MediaPlayerKeyPressed', (n, keyName) => {
+          switch (keyName) {
+            case 'Next': Emitter.sendToGooglePlayMusic('playback:nextTrack'); return;
+            case 'Previous': Emitter.sendToGooglePlayMusic('playback:previousTrack'); return;
+            case 'Play': Emitter.sendToGooglePlayMusic('playback:playPause'); return;
+            case 'Stop': Emitter.sendToGooglePlayMusic('playback:stop'); return;
+            default: return;
+          }
+        });
+        // FIXME: This should be called every time the window is focused
+        //        as this allows multiple applications to share control.
+        iface.GrabMediaPlayerKeys('google-play-music-desktop-player', 0); // eslint-disable-line
+        Logger.info(`Bound media keys for ${service.bus_name}`);
+      }
     }
-  });
+  );
 }
 
 try {
-  const dbus = new DBus();
-  const session = dbus.getBus('session');
+  const session = sessionBus();
 
-  registerBindings('gnome', session);
-  registerBindings('mate', session);
+  const services = [
+    {
+      bus_name: 'org.gnome.SettingsDaemon.MediaKeys',
+      path: '/org/gnome/SettingsDaemon/MediaKeys',
+      interface: 'org.gnome.SettingsDaemon.MediaKeys',
+    },
+    {
+      bus_name: 'org.gnome.SettingsDaemon',
+      path: '/org/gnome/SettingsDaemon/MediaKeys',
+      interface: 'org.gnome.SettingsDaemon.MediaKeys',
+    },
+    {
+      bus_name: 'org.mate.SettingsDaemon',
+      path: '/org/mate/SettingsDaemon/MediaKeys',
+      interface: 'org.mate.SettingsDaemon.MediaKeys',
+    },
+  ];
+
+  Logger.debug('Binding mediakeys');
+  services.forEach((service) => { registerBindings(service, session); });
 } catch (e) {
-  // do nothing.
+  Logger.warn(`Failed to bind mediakeys: ${e}`);
 }
```

**Comment by jostrander** [comment]
The correct move here is probably changing to dbus-native. However we also depend on node-mpris which depends on dbus. As far as I know there is no mpris module that runs on dbus-native, so we would have to integrate it ourselves if we wanted to only depend on one dbus module.

**Comment by calvin-sykes** [comment]
Alternatively I could report this as a bug in dbus?

**Comment by GreatBigWhiteWorld** [comment]
Hello, I'm having exactly this error after a few weeks of use. Today I can't launch the program anymore even after reboot. The error is 
"Segmentation fault (core dumped)" Version rpm 4.6.1.
What do I do?

**Comment by jostrander** [comment]
Probably fixed in #3587
